### PR TITLE
 i#3544 riscv64: Add CI support for RISCV64 

### PR DESCRIPTION
**Pull request by shiptux**
Adds CI support to perform RISCV64 cross compile build. This will
help keep the RISCV64 build healthy by making sure new code
doesn't break anything.

Currently we do not build or run any tests. Added a TODO for that.

Issue #3544

### CODE REVIEW

**Comment by shiptux** [comment]
Could I test the github action workflow in a fork repo?

**Comment by shiptux** [comment]
`https://github.com/DynamoRIO/dynamorio/actions/runs/3741081513/jobs/6350227037`
This CI can currently only compile.

**Comment by derekbruening** [comment]
You can run an action against a branch in the same repo, but only actions already on main show up so a new one can't be tested that way -- but once the file is in place changes to the file can be tested that way.

**Comment by shiptux** [comment]
I have added an github action ci-riscvxx.yml in commit part 2, this same PR.

**Comment by shiptux** [comment]
Those ci-windows workflow are always not pass.Even in other PR.
Another question is the copyright. I saw all of the action files are copyright google. So I keep it consistent. But I don't know if it's right.

**Comment by derekbruening** [comment]
> I had added an github action ci-riscvxx.yml in commit part 2, this same PR.

Dividing into "part 1", "part 2", etc. in the title is for the final merge-to-main commit: we do not use those labels if multiple commits go into the same PR, only for multiple PR's.

**Comment by derekbruening** [comment]
> @[shiptux](https://github.com/shiptux) shiptux [force-pushed](https://github.com/DynamoRIO/dynamorio/compare/134ce476cdb0f29468c527d676838cb4d85d1cd8..fefcc5ee5c46495358c689564e959931a9c36c15) the master branch from [134ce47 ](https://github.com/DynamoRIO/dynamorio/commit/134ce476cdb0f29468c527d676838cb4d85d1cd8)to [fefcc5e ](https://github.com/DynamoRIO/dynamorio/commit/fefcc5ee5c46495358c689564e959931a9c36c15)[yesterday](https://github.com/DynamoRIO/dynamorio/pull/5799#event-8070246369)

Please do not force-push to a shared branch.  It messes up the review history and causes problems if anyone else checked out the branch.  The commit message is cleaned up in the final squash-and-merge so there is no benefit and only negatives to force-pushing a PR.  This is documented in several places: https://dynamorio.org/page_contributing.html, https://dynamorio.org/page_code_reviews.html

**Comment by derekbruening** [comment]
> Those ci-windows workflow are always not pass.Even in other PR.

I think Github updated their Windows VM and caused some problems: someone else hit this in https://github.com/DynamoRIO/dynamorio/pull/5797#issuecomment-1358244930.  One of you should file an issue on this as the first to see it. 

> Another question is the copyright. I saw all of the action files are copyright google. So I keep it consistent. But I don't know if it's right.

Please see https://dynamorio.org/page_contributing.html : "If you would like to contribute code to DynamoRIO, we do not require a formal contributor license agreement or transfer of copyright. Contributions are implicitly assumed to be offered under terms of [DynamoRIO's license](https://dynamorio.org/page_license.html)."

**Comment by derekbruening** [comment]
Please indicate when ready for review.

**Comment by shiptux** [comment]
> > I had added an github action ci-riscvxx.yml in commit part 2, this same PR.
> 
> Dividing into "part 1", "part 2", etc. in the title is for the final merge-to-main commit: we do not use those labels if multiple commits go into the same PR, only for multiple PR's.

I rethought it. That do not need two PR or commit. So i edited the title.

**Comment by shiptux** [comment]
> > @[shiptux](https://github.com/shiptux) shiptux [force-pushed](https://github.com/DynamoRIO/dynamorio/compare/134ce476cdb0f29468c527d676838cb4d85d1cd8..fefcc5ee5c46495358c689564e959931a9c36c15) the master branch from [134ce47 ](https://github.com/DynamoRIO/dynamorio/commit/134ce476cdb0f29468c527d676838cb4d85d1cd8)to [fefcc5e ](https://github.com/DynamoRIO/dynamorio/commit/fefcc5ee5c46495358c689564e959931a9c36c15)[yesterday](https://github.com/DynamoRIO/dynamorio/pull/5799#event-8070246369)
> 
> Please do not force-push to a shared branch. It messes up the review history and causes problems if anyone else checked out the branch. The commit message is cleaned up in the final squash-and-merge so there is no benefit and only negatives to force-pushing a PR. This is documented in several places: https://dynamorio.org/page_contributing.html, https://dynamorio.org/page_code_reviews.html

The last time I did this, to squash two commit. I will not force-push after that.

**Comment by shiptux** [comment]
> Please indicate when ready for review.

i am ready for review

File: suite/runsuite.cmake [hunk 1]
```diff
@@ -546,6 +550,30 @@ if (ARCH_IS_X86 AND UNIX AND (a64_on_x86_only OR NOT arg_automated_ci))
...
 endif ()
 
+if (ARCH_IS_X86 AND UNIX AND cross_riscvxx_linux_only)
+# Do riscv cross-compilation if cross_riscvxx_linux_only is on.
+  set(orig_extra_ctest_args ${extra_ctest_args})
+  set(run_tests OFF)
+  set(prev_optional_cross_compile ${optional_cross_compile})
+  set(prev_run_tests ${run_tests})
+  set(ARCH_IS_X86 OFF)
+  # TODO Because of riscv can't run the build tests. So set it closed here.
```
Comment (derekbruening) [inline suite/runsuite.cmake:560]: The convention is to always have an issue number for a TODO or XXX comment: `TODO i#3544: `

File: suite/runsuite.cmake [hunk 2]
```diff
@@ -546,6 +550,30 @@ if (ARCH_IS_X86 AND UNIX AND (a64_on_x86_only OR NOT arg_automated_ci))
...
+  set(build_tests "BUILD_TESTS:BOOL=OFF")
```
Comment (derekbruening) [inline suite/runsuite.cmake:561]: But the comment just says you can't *run* the tests (hmm actually it says "can't run the build tests": please reword): if you can build them, this option should stay on as we want to test building them.  If you can't build the tests yet, please update the comment.

File: .github/workflows/ci-riscvxx.yml [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+
+  # Manual trigger using the Actions page. May remove when integration complete.
+  workflow_dispatch:
+
+defaults:
+  run:
+    shell: bash
+
+jobs:
+  # RISCV64 cross-compile with gcc, with some tests run under QEMU.
```
Comment (derekbruening) [inline .github/workflows/ci-riscvxx.yml:51]: I don't see any tests run under QEMU in this PR: please update the comment.

File: suite/runsuite.cmake [hunk 4]
```diff
@@ -546,6 +550,30 @@ if (ARCH_IS_X86 AND UNIX AND (a64_on_x86_only OR NOT arg_automated_ci))
     " OFF ${arg_package} "")
 endif ()
 
+if (ARCH_IS_X86 AND UNIX AND cross_riscvxx_linux_only)
+# Do riscv cross-compilation if cross_riscvxx_linux_only is on.
+  set(orig_extra_ctest_args ${extra_ctest_args})
+  set(run_tests OFF)
```
Comment (derekbruening) [inline suite/runsuite.cmake:556]: Please add something like `TODO i#3544: Run tests under QEMU` before this line as the current line is not what we want long-term.

File: suite/runsuite.cmake [hunk 5]
```diff
@@ -546,6 +550,30 @@ if (ARCH_IS_X86 AND UNIX AND (a64_on_x86_only OR NOT arg_automated_ci))
     " OFF ${arg_package} "")
 endif ()
 
+if (ARCH_IS_X86 AND UNIX AND cross_riscvxx_linux_only)
```
Comment (derekbruening) [inline suite/runsuite.cmake:553]: We may want this to behave in the same way as the aarchxx and android cross tests: to be enabled in local suite runs too and not only triggered by this GA CI variable.  The other cross tests will operate locally if the cross compiler is built, so you can run the whole suite on your local machine and include the cross tests.

**Review by derekbruening — COMMENTED** [review]
Thank you for the contribution.  I ended up with a bunch of comments but mostly they are style/nits.  Probably worth another round.  However, I will be traveling and out for a while; maybe @abhinav92003 could do the next round or it's possible I'll be able to at some point in the next week.

File: .github/workflows/ci-riscvxx.yml [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+      with:
+        token: ${{ secrets.GITHUB_TOKEN }}
+      if: ${{ github.event_name == 'pull_request' }}
+
+    - run: git fetch --no-tags --depth=1 origin master
+
+    # Install cross-compiler for cross-compiling Linux build.
+    # Unfortunately there are no libunwind or compression cross-compile
+    # packages so we unpack the native versions and copy their files.
+    # Unfortunately there are only kinetic (ubuntu 22.10) has linunwind-riscv64 packages.
```
[the remark this answers is not in the document]
Comment (shiptux) [reply .github/workflows/ci-riscvxx.yml:73]: I think that deleting "there are" is better here.

File: .github/workflows/ci-riscvxx.yml [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+# DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+# SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+# LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+# OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH
+# DAMAGE.
+
+# Github Actions workflow for RISCVXX Continuous Integration testing.
+
+name: ci-riscvxx
```
[the remark this answers is not in the document]
Comment (shiptux) [reply .github/workflows/ci-riscvxx.yml:33]: The "xx" is cover riscv32 and riscv64. I currently have no plans to support riscv32, but maybe someone else will later.

File: .github/workflows/ci-riscvxx.yml [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+    # packages so we unpack the native versions and copy their files.
+    # Unfortunately there are only kinetic (ubuntu 22.10) has linunwind-riscv64 packages.
+    # I prefer debian-ports here, it is more close to upstream.
+    - name: Create Build Environment
+      run: |
+        sudo apt-get update
+        sudo apt-get -y install doxygen vera++ cmake crossbuild-essential-riscv64 git  \
+        qemu-user qemu-user-binfmt
+        sudo sh -c "echo 'deb [trusted=yes arch=riscv64] http://deb.debian.org/debian-ports sid main' |tee -a /etc/apt/sources.list"
+        sudo apt update
```
[the remark this answers is not in the document]
Comment (shiptux) [reply .github/workflows/ci-riscvxx.yml:81]: "sudo add-apt-repository" does a lot of work including "apt update", if I use it, I don't need to do "sudo apt update" again.

File: .github/workflows/ci-riscvxx.yml [hunk 9]
```diff
@@ -0,0 +1,121 @@
...
+    # Unfortunately there are no libunwind or compression cross-compile
+    # packages so we unpack the native versions and copy their files.
+    # Unfortunately there are only kinetic (ubuntu 22.10) has linunwind-riscv64 packages.
+    # I prefer debian-ports here, it is more close to upstream.
+    - name: Create Build Environment
+      run: |
+        sudo apt-get update
+        sudo apt-get -y install doxygen vera++ cmake crossbuild-essential-riscv64 git  \
+        qemu-user qemu-user-binfmt
+        sudo sh -c "echo 'deb [trusted=yes arch=riscv64] http://deb.debian.org/debian-ports sid main' |tee -a /etc/apt/sources.list"
```
[the remark this answers is not in the document]
Comment (shiptux) [reply .github/workflows/ci-riscvxx.yml:80]: "sudo add-apt-repository" does a lot of work including "apt update", if I use it, I don't need to do "sudo apt update" again.

File: .github/workflows/ci-riscvxx.yml (hunk 3 above)
Comment (shiptux) [reply .github/workflows/ci-riscvxx.yml:51]: Ok. removed that

File: suite/runsuite.cmake (hunk 4 above)
Comment (shiptux) [reply suite/runsuite.cmake:556]: Added it.

File: suite/runsuite.cmake (hunk 2 above)
Comment (shiptux) [reply suite/runsuite.cmake:561]: "#TODO i#3544: Run build tests". Is it better?

File: suite/runsuite.cmake (hunk 1 above)
Comment (shiptux) [reply suite/runsuite.cmake:560]: "#TODO i#3544: Run tests under QEMU"

File: .github/workflows/ci-riscvxx.yml [hunk 10]
```diff
@@ -0,0 +1,121 @@
...
+    # Install cross-compiler for cross-compiling Linux build.
+    # Unfortunately there are no libunwind or compression cross-compile
+    # packages so we unpack the native versions and copy their files.
+    # Unfortunately there are only kinetic (ubuntu 22.10) has linunwind-riscv64 packages.
+    # I prefer debian-ports here, it is more close to upstream.
+    - name: Create Build Environment
+      run: |
+        sudo apt-get update
+        sudo apt-get -y install doxygen vera++ cmake crossbuild-essential-riscv64 git  \
+        qemu-user qemu-user-binfmt
```
[the remark this answers is not in the document]
Comment (shiptux) [reply .github/workflows/ci-riscvxx.yml:79]: Done.

File: .github/workflows/ci-riscvxx.yml (hunk 7 above)
[the remark this answers is not in the document]
Comment (shiptux) [reply .github/workflows/ci-riscvxx.yml:33]: I removed the "xx". It is hard to see riscv32 hardware.

File: suite/runsuite.cmake (hunk 5 above)
Comment (shiptux) [reply suite/runsuite.cmake:553]: Done.

**Comment by shiptux** [comment]
1 Some modifications:
  2 
  3 Fix typos, indent
  4 Rename riscvxx to riscv64
  5 Use "add-apt-repository" instead.
  6 Fix the "TODO" format.

File: suite/runsuite.cmake [hunk 11]
```diff
@@ -319,7 +323,7 @@ endif ()
 # also turned on for release-external-64, but ctest will run with label
 # RUN_IN_RELEASE.
 
-if (NOT cross_aarchxx_linux_only AND NOT cross_android_only AND NOT a64_on_x86_only)
+if (NOT cross_riscvxx_linux_only AND NOT cross_aarchxx_linux_only AND NOT cross_android_only AND NOT a64_on_x86_only)
```
[the remark this answers is not in the document]
Comment (shiptux) [reply suite/runsuite.cmake:326]: Done

File: suite/runsuite.cmake [hunk 12]
```diff
@@ -546,6 +550,30 @@ if (ARCH_IS_X86 AND UNIX AND (a64_on_x86_only OR NOT arg_automated_ci))
     " OFF ${arg_package} "")
 endif ()
 
+if (ARCH_IS_X86 AND UNIX AND cross_riscvxx_linux_only)
+# Do riscv cross-compilation if cross_riscvxx_linux_only is on.
```
[the remark this answers is not in the document]
Comment (shiptux) [reply suite/runsuite.cmake:554]: Done.

File: suite/runsuite.cmake [hunk 13]
```diff
@@ -550,14 +552,17 @@ if (ARCH_IS_X86 AND UNIX AND (a64_on_x86_only OR NOT arg_automated_ci))
     " OFF ${arg_package} "")
 endif ()
 
-if (ARCH_IS_X86 AND UNIX AND cross_riscvxx_linux_only)
-# Do riscv cross-compilation if cross_riscvxx_linux_only is on.
+if (ARCH_IS_X86 AND UNIX)
+  #TODO i#3544: Run tests under QEMU
```
Comment (abhinav92003) [inline suite/runsuite.cmake:556]: Style nit: Add space between the `#` and `TODO`: `# TODO i#3544: ...`

File: suite/runsuite.cmake [hunk 14]
```diff
@@ -550,14 +552,17 @@ if (ARCH_IS_X86 AND UNIX AND (a64_on_x86_only OR NOT arg_automated_ci))
...
   set(orig_extra_ctest_args ${extra_ctest_args})
   set(run_tests OFF)
   set(prev_optional_cross_compile ${optional_cross_compile})
   set(prev_run_tests ${run_tests})
+  if (NOT cross_riscv64_linux_only)
+    set(optional_cross_compile ON)
+  endif ()
   set(ARCH_IS_X86 OFF)
-  # TODO Because of riscv can't run the build tests. So set it closed here.
+  #TODO i#3544: Run build tests.
```
Comment (abhinav92003) [inline suite/runsuite.cmake:565]: Same here. The space between the `#` and `TODO`.

File: suite/runsuite.cmake (hunk 2 above)
Comment (abhinav92003) [reply suite/runsuite.cmake:566]: It'd be clearer to say

`TODO i#3544: Port tests to RISCV and build them in the workflow`

File: suite/runsuite.cmake [hunk 15]
```diff
@@ -421,7 +422,8 @@ if (NOT cross_riscvxx_linux_only AND NOT cross_aarchxx_linux_only AND NOT cross_
         ")
     endif (DO_ALL_BUILDS)
   endif (ARCH_IS_X86 AND NOT APPLE)
-endif (NOT cross_aarchxx_linux_only AND NOT cross_android_only AND NOT a64_on_x86_only)
+endif (NOT cross_riscv64_linux_only NOT cross_aarchxx_linux_only AND
```
Comment (abhinav92003) [inline suite/runsuite.cmake:425]: This is missing an `AND`.

s/NOT cross_riscv64_linux_only NOT/NOT cross_riscv64_linux_only AND NOT/

**Comment by shiptux** [comment]
Are there developer mail lists?

File: suite/runsuite.cmake (hunk 13 above)
Comment (shiptux) [reply suite/runsuite.cmake]: Done.

File: suite/runsuite.cmake (hunk 14 above)
Comment (shiptux) [reply suite/runsuite.cmake]: Done.

File: suite/runsuite.cmake (hunk 2 above)
Comment (shiptux) [reply suite/runsuite.cmake:566]: > It'd be clearer to say
> 
> `TODO i#3544: Port tests to RISCV and build them in the workflow`
Fine.

**Comment by abhinav92003** [comment]
> Are there developer mail lists?

Yes: https://groups.google.com/g/dynamorio-devs.

File: suite/runsuite.cmake (hunk 15 above)
Comment (shiptux) [reply suite/runsuite.cmake:425]: Done. I will push it later.

File: suite/runsuite.cmake [hunk 16]
```diff
@@ -546,6 +552,33 @@ if (ARCH_IS_X86 AND UNIX AND (a64_on_x86_only OR NOT arg_automated_ci))
...
+    ${build_tests}
+    CMAKE_TOOLCHAIN_FILE:PATH=${CTEST_SOURCE_DIRECTORY}/make/toolchain-riscv64.cmake
+    " OFF ${arg_package} "")
+  testbuild_ex("riscv64-release-external" ON "
+    DEBUG:BOOL=OFF
+    INTERNAL:BOOL=OFF
+    ${build_tests}
+    CMAKE_TOOLCHAIN_FILE:PATH=${CTEST_SOURCE_DIRECTORY}/make/toolchain-riscv64.cmake
+    " OFF ${arg_package} "")
+endif ()
```
Comment (abhinav92003) [inline suite/runsuite.cmake:580]: We perhaps also need to restore the optional_cross_compile and run_tests vars (similar to the lines 480 and 482 above).

File: suite/runsuite.cmake [hunk 17]
```diff
@@ -546,6 +552,33 @@ if (ARCH_IS_X86 AND UNIX AND (a64_on_x86_only OR NOT arg_automated_ci))
     " OFF ${arg_package} "")
 endif ()
 
+if (ARCH_IS_X86 AND UNIX)
+  # TODO i#3544: Run tests under QEMU
+  set(orig_extra_ctest_args ${extra_ctest_args})
+  set(run_tests OFF)
```
Comment (abhinav92003) [inline suite/runsuite.cmake:558]: This seems like a bug: we're saving prev_run_tests (at line 560 below) *after* modifying it here. Perhaps this should be moved to below L560?

**Review by abhinav92003 — APPROVED** [review]
Sorry for the fragmented review. Found a couple other things we should fix.

Looks good otherwise. Since you don't have committers access, I'll merge the PR after the comments are resolved.

**Comment by abhinav92003** [comment]
The `client.file_io` test is failing on vs2019-32: it's #5802.

**Comment by abhinav92003** [comment]
Two CI workflows are failing. Both due to #5802. Should be safe to merge this PR since it is completely independent.

**Comment by shiptux** [comment]
> > > > Are there developer mail lists?
> > > 
> > > 
> > > Yes: https://groups.google.com/g/dynamorio-devs.
> > 
> > 
> > Could I join this list?
> 
> Of course! Please send a request to join the list on the above link. I don't seem to have admin access to that group, so @derekbruening would need to approve it when he's back.

My email is shiptux@gmail.com.  Thanks.

**Comment by shiptux** [comment]
> > > > > > Are there developer mail lists?
> > > > > 
> > > > > 
> > > > > Yes: https://groups.google.com/g/dynamorio-devs.
> > > > 
> > > > 
> > > > Could I join this list?
> > > 
> > > 
> > > Of course! Please send a request to join the list on the above link. I don't seem to have admin access to that group, so @derekbruening would need to approve it when he's back.
> > 
> > 
> > My email is [shiptux@gmail.com](mailto:shiptux@gmail.com). Thanks.
> 
> I've sent you an invite to join the list.

Sorry for saw it too late, The Github mention send as other mail address. link was expired. Could you please invite me again, Thanks.

**Pull request merged**